### PR TITLE
improve builder performance in @babel/types

### PR DESCRIPTION
**Pull request by zxbodya**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

This is experimental try to improve node builders in @babel/types. Feedback and suggestions are welcome.

Goals:

- try generating complete builders instead of dynamically building nodes - having juts one object literal might be faster
- experiment about removing node validation from builders -- idea is that this is probably useful only during development, but if typescript is used - this should be already covered for most cases… so - it sounds like a good candidate to remove/disable in production

Todo:

- come-up with realistic approach to measure how faster it can be if generated and how much overhead there is because of validation -- suggestions and help is welcome

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13369"></a>

### CODE REVIEW

File: packages/babel-types/scripts/generators/builders.js [hunk 1]
```diff
@@ -96,11 +96,47 @@ import type * as t from "../..";
...
     output += `${
       formatedBuilderNameLocal === formatedBuilderName ? "export " : ""
     }function ${formatedBuilderNameLocal}(${defArgs.join(
       ", "
-    )}): t.${type} { return builder("${type}", ...arguments); }\n`;
+    )}) {\n  const node = {\n${objectFields
+      .map(([k, v]) => `    ${k}: ${v},`)
+      .join(
+        "\n"
+      )}  } as t.${type};\n  validateNode(node);\n  return node;\n}\n`;
```
Comment (zxbodya) [inline packages/babel-types/scripts/generators/builders.js:138]: thinking if this `validateNode(…)` call can be removed…

Comment (JLHwung) [reply packages/babel-types/scripts/generators/builders.js:138]: I think we have customized validation that can only be invoked via `validateNode`.

File: packages/babel-types/src/builders/generated/index.ts [hunk 2]
```diff
@@ -45,179 +57,374 @@ export function binaryExpression(
...
 }
 export function catchClause(
   param: t.Identifier | t.ArrayPattern | t.ObjectPattern | null | undefined,
   body: t.BlockStatement,
-): t.CatchClause {
-  return builder("CatchClause", ...arguments);
+) {
+  const node = {
+    type: "CatchClause",
+    param: param === undefined ? null : param,
```
Comment (JLHwung) [inline packages/babel-types/src/builders/generated/index.ts:137]: Can we replace `# === undefined ? a : b` by default parameters, if it is one of the builder keys.

**Review by JLHwung — COMMENTED** [review]
On the performance benchmark, we have micro benchmark cases in `babel-parser`: `packages/babel-parser/benchmark`

File: packages/babel-types/src/builders/generated/index.ts (hunk 2 above)
Comment (zxbodya) [reply packages/babel-types/src/builders/generated/index.ts:137]: yeah, actually already tried that - let me push it

**Comment by JLHwung** [comment]
The benchmark is for general purpose. For example I have drafted a benchmark on the string literal builder: https://github.com/JLHwung/babel/commit/44e6253dcabda15c7b0d8b156140ac498deb3b8f

```
node --predictable ./builder-string-literal/bench.mjs
baseline stringLiteral builder: 3764552 ops/sec ±1.06% (0ms)
current stringLiteral builder: 5278258 ops/sec ±0.77% (0ms)
```

The new implementation is 40% faster.

**Pull request closed without merging**